<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>callback and promise</title>
</head>
<body>
<h1 style="text-align:center;">CallBack and Promise Function</h1>
<h2>Question1- <br> Write one example how you can write a callback function</h2>
<pre>
    function demo(myfun){
        console.log('hello');
        myfun();
        }
        function demo2(){
        console.log('saya')
        }
        demo(demo2);
</pre>
<p>The demo function logs the message "hello" to the console, and then calls the myfun function, which in this case is demo2. <br>The demo2 function logs the message "saya" to the console.</p>

<h2>Question2- <br> Write a callback function to print numbers from 1 to 7, in which 1 should be printed after 1 sec,<br> 2 should be printed after 2 sec, 3 should be printed after 3 sec, and so on. 
<pre>
    Explain callback hell.
    Numbers
    1
    2
    3
    4
    5
    6
    7
</pre></h2>
<pre>
    function callback(){
        for(let i=1 ; i <=7 ; i++){
        
            setTimeout(()=>{
                console.log(i);
            },i*1000)
        }
    }
    callback();
</pre>
<p>when callback function is called, for loop will run from 1 to 7, and prints the number after<br> every one second. Here when loop runs setTimeout() executes by having an anonymous function which prints the number after every 1sec. </p>
<p>The phenomenon which happens when we nest multiple callbacks within a function is called a callback hell. The shape of the resulting code structure<br> resembles a pyramid and hence callback hell is also called the “pyramid of the doom”. It makes the code very difficult to understand and maintain.</p>


<h2>Question3- <br> Write the promise function to print numbers from 1 to 7, in which 1 should be printed after 1 sec, <br>2 should be printed after 2 sec, 3 should be printed after 3 sec, and so on. 
    <pre>
   Numbers
    1
    2
    3
    4
    5
    6
    7
    </pre></h2>
    <pre>

        function print(num){
            return new Promise((resolve)=>{
                setTimeout(()=>{
                     console.log(num);
                     resolve()
                 },1000)
             })
          }
          print(1) 
          .then(()=>print(2))
          .then(()=>print(3))
          .then(()=>print(4))
          .then(()=>print(5))
          .then(()=>print(6))
          .then(()=>print(7))


    </pre>
    <p>when print function is called, which have num as an argument, returns promise. promise returned by print function have setTimeout method which prints the num after a delay of 1sec.<br>For resolve method .then method is used , so here .then method is used to print the number.  </p>

    <h2>Question4-Create a promise function accepting an argument, if yes is passed to the function then it should go to <br>resolved state and print Promise Resolved, and if nothing is passed then it should go to reject the state and <br>catch the error and print Promise Rejected</h2>

    <pre>
        function promise(argu) {
            return new Promise((resolve, reject) => {
              if (arg === 'yes') {
                resolve('Promise Resolved');
              } else {
                reject('Promise Rejected');
              }
            });
        }
          
        promise('yes')

        .then((result) => {
            console.log(result);
        })
        .catch((error) => {
            console.log(error);
        });
    </pre>
    <p>here  promise function accepts the argument argu, it returns a promise with the parameters resolve and reject.<br> if the argu is yes, resolve method will print promise resolved, otherwise reject method will print promise rejected. when promise is called with "yes" as parameter, <br>then the .then method with result as an argument will print the resolve method and .catch will print the reject method.</p>

    <h2>Question5-Create examples to explain callback function</h2>
    <pre>
        function demo(myfun){
            console.log("hey");
            myfun();
        }

        function demo2(){
            console.log("saya");
        }

        demo(demo2);
    </pre>
    <p>when demo function is called with demo2 as a parameter, the function declaration of demo having myfun<br> as an argument which is called inside the function is a callback function. so, first it will print "hey" then since demo2 is a paramter called in demo, <br>the demo2 function will run and print "saya" </p>
    
    <h2>Question6-Create examples to explain callback hell function</h2>
    <pre>
        function higher1(callback){
            console.log("hello1");
            callback();
        }
        function higher2(callback){
            console.log("hello2");
            callback();
        }
        function higher3(){
            console.log("hello3");
        }

        higher1(function(){
            higher2(function(){
                higher3();
            })
        })
    </pre>

    <p>when higher1 function is called, higher1 function with argument callback will print hello1, <br>the argument callback is called inside the higher1 function, which makes callback a callback function. similarly higher2. when higher3 is called , hello3 will be printed. <br>Calling all these function, means function inside function is an example of callback hell function. </p>

    <h2>Question7-Create examples to explain promise function</h2>
    <pre>
        function promise(argu){
            return new Promise((resolve, reject)=>{
                
                if(argu="yes"){
                    resolve("promise resolved");
                }
                else{
                    reject("promise rejected");
                }
            })
        }

        promise("yes").then((result)=>{
            console.log(result);
        })
        .catch((error)=>{
            console.log(error);
        })
    </pre>
    <p>here  promise function accepts the argument argu, it returns a promise with the parameters resolve and reject.<br> if the argu is yes, resolve method will print promise resolved, otherwise reject method will print promise rejected. when promise is called with "yes" as parameter, <br>then the .then method with result as an argument will print the resolve method and .catch will print the reject method.</p>



    <h2>Question8-Create examples to explain async await function</h2>
    <pre>
        let demo = new Promise((resolve)=>{
            setTimeout(()=>{
                resolve("promise resolved")
            },2000)
        })

        async function sample(){
            let a = await demo;
            console.log(a);
            console.log("this is async method");
        }

        sample();
    </pre>
    <p>here a promise is getting created with resolve method, which is stored in demo, which performs setTimeout function,to result after 2secs. when sample function is called,<br>flow of control knows that sample function is an asynchronous function which comes with await keyword. await will make sure that after demo , sample function will execute with the delay of 2secs. </p>


    <h2>Question9-Create examples to explain Promise.all function</h2>
    <pre>
        const promise1 = Promise.resolve(24);
        const promise2 = 48;
        const promise3 = new Promise((resolve, reject) => {
        setTimeout(resolve, 1000, 'foot');
        });

        Promise.all([promise1, promise2, promise3]).then((values) => {
        console.log(values);
        });
    </pre>
    <p>Promise.all() is a method which takes an array of promises and results in a single promise. This returned promise fulfills when all of the input's promises fullfill. Promise.all() rejects when<br> any of the input's promise rejects. <br> 
        promise1, promise2, promise3 are three promises. promise1 have a promise reoslve method with results in 24, promise2 just have a number 48,promise3 is creatinga promise with <br>setTimeout function that resolves 24 after  1sec, with foot as a string. with promise.all method the result will be in the form of array with promise1, promise2,promise3 promises. <br>prints the output in the same order when (promise) is passed</p>
</body>
</html>